### PR TITLE
chore: open-source our git_push rule

### PR DESCRIPTION
**Pull request by alexeagle**

### CODE REVIEW

File: git_push/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# git_push example
+
+This shows how you can make an executable rule that publishes a subset of `bazel-out` to a GitHub repository.
+
+This is useful when your monorepo wants to export some package, for example an SDK for your APIs.
+
+Aspect uses this ourselves to publish our workflows packages from our internal monorepo, for example https://github.com/aspect-build/workflows-action comes from one folder in our repo.
```
Comment (gregmagolan) [inline git_push/README.md:7]: nit: workflows with a capital W

File: git_push/defs.bzl [hunk 2]
```diff
@@ -0,0 +1,39 @@
+"git_push rule for delivering to GitHub"
+
+# We can't use the bazel-lib one, because it doesn't have a program to read stamp vars.
+# see https://github.com/aspect-build/rules_js/pull/384#issue-1337742941
+# load("@aspect_bazel_lib//lib:expand_make_vars.bzl", "expand_template")
+# buildifier: disable=bzl-visibility
+load("@aspect_rules_js//js/private:expand_template.bzl", "expand_template")
```
Comment (gregmagolan) [inline git_push/defs.bzl:7]: lol. my little private utility is getting a lot of public attention

File: git_push/defs.bzl [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+def git_push(name, archive, repo_url, **kwargs):
+    """Push a commit to the given git remote, replacing the repo content with the content of the archive.
+
+    Args:
+        name: name of resulting sh_binary target
```
Comment (gregmagolan) [inline git_push/defs.bzl:13]: nit:
```
name: unique name for the runnable sh_binary target
```
?

File: git_push/defs.bzl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+# We can't use the bazel-lib one, because it doesn't have a program to read stamp vars.
+# see https://github.com/aspect-build/rules_js/pull/384#issue-1337742941
+# load("@aspect_bazel_lib//lib:expand_make_vars.bzl", "expand_template")
+# buildifier: disable=bzl-visibility
+load("@aspect_rules_js//js/private:expand_template.bzl", "expand_template")
+
+def git_push(name, archive, repo_url, **kwargs):
+    """Push a commit to the given git remote, replacing the repo content with the content of the archive.
+
```
Comment (gregmagolan) [inline git_push/defs.bzl:11]: nit: mention default tags are `["deliverable"]` if tags are not set

File: git_push/push_archive_to_repo.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
+# Executable that unpacks an archive, and pushes the content as a git repo
+# Assumes that `git` is on the machine and has push credentials already.
+# Inspired from how Angular publishes snapshots to "*-build" repos:
```
Comment (gregmagolan) [inline git_push/push_archive_to_repo.sh:4]: nit: also assumes that `tar` is on the machine... it should be on all major linux distros and all on modern Windows releases

Comment (gregmagolan) [reply git_push/push_archive_to_repo.sh:4]: mention that `MYORG_GHBOT_PAT` needs to be renamed and set as well?

Comment (gregmagolan) [reply git_push/push_archive_to_repo.sh:4]: nit: also assumes `mktemp` is on the machine

File: git_push/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        "README.md",
+        "some.txt",
+    ],
+)
+
+git_push(
+    name = "release",
+    archive = "assets.tar",
+    repo_url = "https://github.com/my-org/workflows-action.git",
+)
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply git_push/BUILD.bazel:24]: so analysis is checked at least and it doesn't bit rot

**Review by gregmagolan — APPROVED** [review]

File: git_push/defs.bzl (hunk 2 above)
Comment (alexeagle) [reply git_push/defs.bzl:7]: yes we really need to write another tiny go program with a toolchain hanging off bazel-lib so we can upstream that

File: git_push/defs.bzl (hunk 4 above)
Comment (alexeagle) [reply git_push/defs.bzl:11]: actually the deliverable tagging scheme is orthogonal to the changes going on here

**Pull request merged**